flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/optimize/program/OptimizeContext.scala

```diff
@@ -0,0 +1,37 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.plan.optimize.program
+
+import org.apache.calcite.plan.{Context, RelOptPlanner}
+
+/**
+  * A OptimizeContext allows to obtain table environment information when optimizing.
+  */
+trait OptimizeContext {
+
+  /**
+    * Gets the Calcite [[Context]] defined in [[org.apache.flink.table.api.TableEnvironment]].
+    */
+  def getContext: Context
```

Comment:
Could we try to avoid use Calcite's Context to pass some necessary staffs? The Calcite's Context is really obscure to tell what exactly we can get from it. In this case, you might thought use this pass `TableConfig` around, but no one can know this though this API.

Reply:
Only `TableConfig` is not enough, `FunctionCatalog` is also needed by `RexNodeExtractor` which is used in `PushFilterIntoTableSourceScanRule` and `SelectivityEstimator` in Blink.

Reply:
Yes, what i meant is why not just show these two method to OptimizeContext, like
`getTableConfig` and `getFunctionCatalog`. Through `Context`'s API, nobody knows what they can get through it.

Reply:
For most of program, they need not to know the content in the `Context`. The `Context` instance is only be used to create `HepPlanner` now as a whole instance: `new HepPlanner(hepProgram, context.getContext)`.  Calcite's `Context` is built in `TableEnvironment` now, `VolcanoPlanner` and `HepPlanner` are built with this context instance, however they are built in difference place, `VolcanoPlanner` is built in `FlinkRelBuilder` and `HepPlanner` is built in `FlinkHepProgram`. Someone may be forget to modify `FlinkHepProgram` when he/she wants to extend `Context` in `TableEnvironment`.

Reply:
If you don't depend on `Context` to pass things like `TableConfig` around, is it still necessary to build `Context` in `TableEnvironment`? I noticed `HepPlanner` has a constructor without `Context`.

Reply:
A empty `Context` will be created in `AbstractRelOptPlanner` if building a `HepPlanner` without `Context`. And the rules executed in `HepPlanner` can not get `TableConfig` or  `
FunctionCatalog` instances set by user.

Reply:
Got your point. What do you think about we just let `OptimizeContext` to inherit from Calcite's `Context`, and exposes some meaningful APIs to `OptimizeContext`. And in each rule which wants to use `TableConfig`, just get `Context` and cast it to OptimizeContext. 

Another comment on `BatchOptimizeContext` and `StreamOptimizeContext`, can we also try to unify these two? I know there are some differences between them in Blink's code, could you consider possibility of it? Since we may unify table environments for both batch and streaming.

Reply:
Yes, it's more clear than before that `OptimizeContext` extends Calcite's `Context` and exposes meaningful APIs.

`BatchOptimizeContext` and `StreamOptimizeContext` can be unified after table environments have been unified. We can do this later.